File: mqtt-client-Android/app/src/main/java/io/emqx/mqtt/MessageFragment.java
package io.emqx.mqtt;

import android.view.View;

import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;


public class MessageFragment extends BaseFragment {


    private MessageRecyclerViewAdapter mAdapter;

    private List<Message> mList = new ArrayList<>();

    public MessageFragment() {
    }


    public static MessageFragment newInstance() {
        return new MessageFragment();
    }

    @Override
    protected int getLayoutResId() {
        return R.layout.fragment_message_list;
    }

    @Override
    protected void setUpView(View view) {
        RecyclerView recyclerView = view.findViewById(R.id.message_list);
        recyclerView.addItemDecoration(new DividerItemDecoration(fragmentActivity, DividerItemDecoration.VERTICAL));
        mAdapter = new MessageRecyclerViewAdapter(mList);
        recyclerView.setAdapter(mAdapter);

    }


    public void updateMessage(Message message) {
        mList.add(0, message);
        mAdapter.notifyItemInserted(0);

    }


}

File: mqtt-client-Android-Kotlin/app/src/main/java/io/emqx/mqtt/MessageFragment.kt
package io.emqx.mqtt

import android.view.View
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.RecyclerView

class MessageFragment : BaseFragment() {
    private var mAdapter: MessageRecyclerViewAdapter? = null
    private val mList: ArrayList<Message> = ArrayList()


    override val layoutResId: Int
        get() = R.layout.fragment_message_list

    override fun setUpView(view: View) {
        val recyclerView = view.findViewById<RecyclerView>(R.id.message_list)
        recyclerView.addItemDecoration(
            DividerItemDecoration(
                fragmentActivity,
                DividerItemDecoration.VERTICAL
            )
        )
        mAdapter = MessageRecyclerViewAdapter(mList)
        recyclerView.adapter = mAdapter
    }

    fun updateMessage(message: Message) {
        mList.add(0, message)
        mAdapter!!.notifyItemInserted(0)
    }

    companion object {
        fun newInstance(): MessageFragment {
            return MessageFragment()
        }
    }
}